docs/OpenSSLStrategicArchitecture.md

```diff
@@ -0,0 +1,290 @@
+---
+title: OpenSSL Strategic Architecture
+author: OpenSSL Management Committee (OMC)
+date: January, 2019
+---
+## Introduction
+
+This document outlines the OpenSSL strategic architecture. It will take
+multiple releases, starting from 3.0.0, to move the architecture from
+the current "as-is" (1.1.1), to the future "to-be" architecture.
+
+Numerous changes are anticipated in the to-be architecture. A migration
+path for handling the eventual transition will be provided. The OpenSSL
+3.0.0 release will have minimal impact to the vast majority of existing
+applications, almost all well-behaved applications will just need to be
+recompiled.
+
+The current functionality provided by the engine interface will be
+replaced over time via a provider interface. OpenSSL 3.0.0 will continue
+to support engines. The to-be architecture will not be fully realised
+until OpenSSL 4.0.0 at the earliest.
+
+## As-is architecture
+
+Currently, OpenSSL is split into four principal components:
+
+1.  libcrypto. This is the core library for providing implementations of
+    numerous cryptographic primitives. In addition it provides a set of
+    supporting services which are used by libssl and libcrypto, as well
+    as implementations of protocols such as CMS and OCSP.
+
+2.  Engine. The functionality of libcrypto can be extended through the
+    Engine API.
+
+    Typically engines are dynamically loadable modules that are registered
+    with libcrypto and use the available hooks to provide cryptographic
+    algorithm implementations. Usually these are alternative implementations
+    of algorithms already provided by libcrypto (e.g. to enable hardware
+    acceleration of the algorithm), but they may also include algorithms not
+    implemented in default OpenSSL (e.g. the afalg engine provides access to
+    the linux kernel cryptographic algorithms). Some engines are provided as
```

Comment:
I just noticed that this is a bad example. All afalg implementations are for algorithms also available in libcrypto! GOST is a better (perhaps the only) example.

Reply:
I ended up refering to GOST here too, 'cause yeah